src/core/internal/spinlock.d

```diff
@@ -0,0 +1,99 @@
+/**
+ * SpinLock for runtime internal usage.
+ *
+ * Copyright: Copyright Digital Mars 2015 -.
+ * License:   $(WEB www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+ * Authors:   Martin Nowak
+ * Source: $(DRUNTIMESRC core/internal/_spinlock.d)
+ */
+module core.internal.spinlock;
+
+import core.atomic, core.thread;
+
+shared struct SpinLock
+{
+    /// for how long is the lock usually contended
+    enum Contention : ubyte
+    {
+        brief,
+        medium,
+        lengthy,
+    }
+
+@trusted @nogc nothrow:
+    this(Contention cont)
+    {
+        this.cont = cont;
+    }
+
+    void lock()
+    {
+        // TTAS lock (https://en.wikipedia.org/wiki/Test_and_test-and-set)
+        while (true)
+        {
+            for (size_t n; atomicLoad!(MemoryOrder.raw)(val); ++n)
+                yield(n);
+            if (cas(&val, false, true))
+                return;
+        }
+    }
+
+    void unlock()
+    {
+        atomicStore!(MemoryOrder.rel)(val, false);
+    }
+
+    /// yield with backoff
+    void yield(size_t k)
+    {
+        if (k < pauseThresh >> cont)
+            return pause();
+        else if (k < 32 >> cont)
+            return Thread.yield();
+        Thread.sleep(1.msecs);
+    }
+
+private:
+    version (D_InlineAsm_X86)
+        enum X86 = true;
+    else version (D_InlineAsm_X86_64)
+        enum X86 = true;
+    else
+        enum X86 = false;
+
+    static if (X86)
+    {
+        enum pauseThresh = 16;
+        void pause()
+        {
+            asm @trusted @nogc nothrow
+            {
+                // pause instruction
+                rep;
+                nop;
+            }
+        }
+    }
+    else
+    {
+        enum pauseThresh = 4;
+        void pause()
+        {
+        }
+    }
+
+    bool val;
```

Comment:
Have you tried other data sizes? It's possible that 32-bit may be nicer for the CPU.